gprofiler/profilers/node.py

```diff
@@ -0,0 +1,244 @@
+    # ref: https://github.com/nodejs/node/blob/2849283c4cebbfbf523cc24303941dc36df9332f/src/node_options.h#L90
+    # in our case it won't be changed
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
+    if debugger_url_response.status_code != 200 or "application/json" not in debugger_url_response.headers.get(
+        "Content-Type", ""
+    ):
+        raise NodeDebuggerUrlNotFound(
+            {"status_code": debugger_url_response.status_code, "text": debugger_url_response.text}
+        )
+
+    response_json = debugger_url_response.json()
+    if (
+        not isinstance(response_json, list)
+        or len(response_json) == 0
+        or not isinstance(response_json[0], dict)
+        or "webSocketDebuggerUrl" not in response_json[0]
+    ):
+        raise NodeDebuggerUrlNotFound(response_json)
+
+    return cast(str, response_json[0]["webSocketDebuggerUrl"])
+
+
+def _send_socket_request(sock: WebSocket, cdp_request: Dict) -> None:
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
+    ):
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _get_node_pid(sock: WebSocket) -> int:
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": "process.pid",
+            "replMode": True,
+        },
+    }
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+    try:
+        return cast(int, message["result"]["result"]["value"])
+    except KeyError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+
+def _load_dso(sock: WebSocket, module_path: str) -> None:
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": f'process.mainModule.require("{os.path.join(module_path, "linux-perf.js")}").start()',
+            "replMode": True,
+        },
+    }
+    _send_socket_request(sock, cdp_request)
+
+
+def _stop_dso(sock: WebSocket, module_path: str) -> None:
+    cdp_request = {
+        "id": 1,
+        "method": "Runtime.evaluate",
+        "params": {
+            "expression": f'process.mainModule.require("{os.path.join(module_path, "linux-perf.js")}").stop()',
+            "replMode": True,
+        },
+    }
+    _send_socket_request(sock, cdp_request)
+
+
+def validate_ns(nspid: int, expected_ns_link_name: str) -> None:
+    actual_nspid_name = os.readlink(os.path.join("/proc", str(nspid), "ns", "pid"))
```

Comment:
Continuing our discussion from https://github.com/Granulate/gprofiler/pull/451#discussion_r984097131 -

This validation is definitely an improvement - it catches the case where multiple node processes are running on the host, and multiple node processes are running in the same container.

It doesn't catch, however, these 2 cases:
* 2 node processes running in 2 different containers in the same k8s Pod - in k8s, the network namespace is shared between containers in the same Pod, but the PID namespace is not.
* 2 node processes running in 2 different containers with `--net=host`.
These 2 cases are much more exotic, I'd say, but if it doesn't add too much complexity then we can support them as well. For that - it'd require to run this logic `os.readlink(os.path.join("/proc", str(nspid), "ns", "pid"))` via `Runtime.evaluate` (i.e, it's JS equivalent).

I'm okay with leaving it for a future PR though - so you can try it out, and if it turns out not straight forward, we can leave a TODO in code (can refer to the link of thise comment) & I'll open a ticket for later.

Reply:
Implemented, resolving this